docs/localizing.rst

```diff
@@ -403,6 +403,95 @@ Finally, we update the line with our modified line using the **update_line** met
 
                 oFile.lines[iLineNumber].update_line(sNewLine)
 
+Violation dictionary
+--------------------
+
+Violations are stored as a list of dictionaries in the **rule.violations** attribute.
+This is the generic format of the dictionary represented by json:
+
+.. code-block:: json
+
+   {
+     "lines" : [ 
+        {
+          "number" : "<integer>",
+          "<line_attribute>" : "<line_value>",
+          "<line_attribute>" : "<line_value>"
+        }
+       ],
+     "<violation_attribute>" : "<violation_value>",
+     "<violation_attribute>" : "<violation_value>"
+   }
```

Comment:
Maybe we should keep `lines` as a pure integer array with line numbers. not an array of dictionaries? If we need more data associated with given line we could always add another list or lists and associate data by index (lists are ordered).
This would make `lines` attribute more clear for rules that need only line numbers. No need to have `number` key.

Reply:
what I like about the format is it keeps the information atomic.  So information about the line is kept together and information about all the lines are kept together.

The keyword alignment rule would create a violation dictionary like this:

```json
{
  "lines" : [
    {
      "number" : 31,
      "keyword_column" : 52,
      "before_keyword_column" : 12
    },
    {
      "number" : 32,
      "keyword_column" : 45,
      "before_keyword_column" : 6
    },
    {
      "number" : 32,
      "keyword_column" : 76,
      "before_keyword_column" : 30
    }
  ],
  "max_keyword_column" : 76,
  "max_keyword_before_column" : 30
}
```
We can always write a function to pull out all the line numbers:

```python
def extract_line_numbers_from_violation_dict(dViolation):
    lReturn = []
    for dLine in dViolation['lines']:
        lReturn.append(dLine['number'])
    return lReturn
```
it seems like overkill for single line violations, but I have started using these functions:

```python
def create_violation_dict(iLineNumber):
    '''
    Builds a minimal violation dictionary.

    Parameters:

      iLineNumber: (integer)

    Returns:  dictionary
    '''
    dReturn = {}
    dReturn['lines'] = []
    dLine = {}
    dLine['number'] = iLineNumber
    dReturn['lines'].append(dLine)
    return dReturn


def get_violation_linenumber(dViolation):
    '''
    Returns a line number of a violation.

    Parameters:

      dViolation: Violation dictionary

    Returns:  integer
    '''
    return dViolation['lines'][0]['number']


def  get_violating_line(oFile, dViolation):
     '''
     Returns a line from the file where a violation has occured.

     Parameters:

       oFile : (File object)

       dViolation : (Violation dictionary)

     Return: Line Object
     '''
     return oFile.lines[get_violation_linenumber(dViolation)]
```
To help abstract away the structure of the dictionary.

Reply:
Well, no matter which way we choose, it would be more coherent than what we have right now.